Python Wrapper for SystemDS

kev-inn opened the pull request:
We can create matrices from files or numpy. Basic arithmetic is implemented and tested, as well as some aggregations.

Federated works, but I have not added tests because we have to start a worker and we might want to discuss how we distribute the systemds jar. Currently either a standalone jar or our SystemDS jar + classes and libs has to placed in the directory `systemds-java` for the implementation to work, or the java jmlc gateway is manually started and the method `set_aut_start_jmlc(False)` is called.

I did not yet have a detailed look into packaging and releasing on pip, but it seems easy with `setuputils`.

Currently on top of todo list are test cases for federated.

For comparisons like `<`, `>=` etc. we also have to talk about either giving a `DAGNode` back on which `.compute()` has to be called, or calling `.compute()` for both parts and comparing the results. Further discussions which actions except `.compute()` actually start the execution in SystemDS is necessary.

kev-inn committed:
Add tests, binary operations and federated matrices

We can create matrices from files with `.mtd` or numpy. Basic arithmetic is implemented and tested, as well as some aggregations.

Federated works, but I have not added tests because we have to start a worker and we might want to discuss how we distribute the systemds jar. Currently either a standalone jar or our SystemDS jar + classes and libs has to placed in the directory `systemds-java` for the implementation to work, or the java jmlc gateway is manually started and the method `set_aut_start_jmlc(False)` is called.

I did not yet have a detailed look into packaging and releasing on pip, but it seems easy with `setuputils`.

Currently on top of todo list are test cases for federated.

For comparisons like `<`, `>=` etc. we also have to talk about either giving a `DAGNode` back on which `.compute()` has to be called, or calling `.compute()` for both parts and comparing the results. Further discussions which actions except `.compute()` actually start the execution in SystemDS is necessary.

kev-inn committed:
Update licenses

kev-inn commented:
The failing test case (`Build Failed`) works locally, I will take a look into it.

@Baunsgaard recommended rewriting the test cases to check against DML-Script we expect to be created and not execute it with JMLC Gateway server, therefore we would not have to build the SystemDS jar for the test cases.

Baunsgaard commented:
The build fails because of one of my tests, it should be fixed if you merge with master again.

Baunsgaard commented on src/main/java/org/tugraz/sysds/pythonapi/pythonDMLScript.java:
```diff
@@ -0,0 +1,36 @@
+			try {
+				BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
+				stdin.readLine();
+				System.exit(0);
```
If you split this up, such that you make a public function that returns a boolean for if the program fails or passes, then the testing is easier. If you use System exit 0 and 1 then it also will exit the testing.

So i would suggest moving this main methods logic to another function, and then do the System exit here in the main function, while returning a Boolean from the new function.

Baunsgaard commented on src/main/java/org/tugraz/sysds/runtime/instructions/spark/utils/RDDConverterUtilsExt.java:
```diff
@@ -103,11 +103,19 @@
 
 	public static MatrixBlock convertPy4JArrayToMB(byte [] data, long rlen, long clen) {
-		return convertPy4JArrayToMB(data, (int)rlen, (int)clen, false);
+		return convertPy4JArrayToMB(data, (int)rlen, (int)clen, false, 2);
```
why cast? can the inputs not be Ints per default?

kev-inn replied:
I removed the long methods for now, because I agree they should never get used by py4j at the moment.

Baunsgaard commented on src/main/python/setup.py:
```diff
@@ -0,0 +1,97 @@
+    ''',
+    url='https://github.com/tugraz-isds/systemds',
+    author='SystemDS',
+    author_email='dev@systemds.apache.org',
```
is / will this be correct?

Baunsgaard commented on src/main/python/setup.py:
```diff
@@ -0,0 +1,97 @@
+    SystemDS is a versatile system for the end-to-end data science lifecycle from data integration, 
+    cleaning, and feature engineering, over efficient, local and distributed ML model training, 
+    to deployment and serving. To this end, we aim to provide a stack of declarative languages with
+    R-like syntax for 
```
Maybe not mention R?

Suggested change:

SystemDS is a versatile system for the end-to-end data science lifecycle from data integration, 
    cleaning, and feature engineering, over efficient, local and distributed ML model training, 
    to deployment and serving.
To facilitate this, bindings from different languages and different system abstractions provide help for:

Baunsgaard commented on src/main/python/setup.py:
```diff
@@ -0,0 +1,97 @@
+    R-like syntax for 
+    (1) the different tasks of the data-science lifecycle, and 
+    (2) users with different expertise. 
+    These high-level scripts are compiled into hybrid execution plans of local, in-memory CPU and GPU operations, 
```
Newline!

Baunsgaard commented on src/main/python/systemds/converters.py:
```diff
@@ -0,0 +1,46 @@
+    # data_type: 0: int, 1: float and 2: double
+    if np_arr.dtype is np.dtype(np.int32):
+        arr = np_arr.ravel().astype(np.int32)
+        data_type = 0
```
we might want to consider something else than marking it as a integer for each datatype.
There should be enum types in python as i remember.
The other option is to use the already defined types from NumPY.

Baunsgaard commented on src/main/python/systemds/matrix.py:
```diff
@@ -0,0 +1,127 @@
+
+
+# TODO consider different naming for functions
+def full(shape: Tuple[int, int], value: Union[float, int]) -> OperationNode:
```
call it dense Matrix maybe?

Baunsgaard commented on src/main/python/systemds/matrix.py:
```diff
@@ -0,0 +1,127 @@
+    return OperationNode('matrix', unnamed_input_nodes, named_input_nodes)
+
+
+def seq(start: Union[float, int], stop: Union[float, int] = None, step: Union[float, int] = 1) -> OperationNode:
```
Call it Vector?

Baunsgaard replied:
like the numpy seq

Baunsgaard commented on src/main/python/systemds/operation_node.py:
```diff
@@ -0,0 +1,218 @@
+        return OperationNode('abs', [self])
+
+    def moment(self, moment, weights: DAGNode = None) -> OperationNode:
+        # TODO write tests
```
why is this comment only here?

kev-inn replied:
because this function is the only one that is still missing tests.

Baunsgaard commented on src/main/java/org/tugraz/sysds/pythonapi/pythonDMLScript.java:
```diff
@@ -0,0 +1,36 @@
+package org.tugraz.sysds.pythonapi;
```
missing license

Baunsgaard commented on src/main/java/org/tugraz/sysds/runtime/instructions/spark/utils/RDDConverterUtilsExt.java:
```diff
@@ -103,11 +103,19 @@
 	}
 
 	public static MatrixBlock convertPy4JArrayToMB(byte [] data, long rlen, long clen) {
-		return convertPy4JArrayToMB(data, (int)rlen, (int)clen, false);
```
Update License

Baunsgaard commented on src/main/resources/python/project_info.py:
```diff
@@ -0,0 +1,27 @@
+#-------------------------------------------------------------
+#
```
Update to Modification License

Baunsgaard reviewed (commented):
LGTM,  :100: :1st_place_medal: :+1:   
only cosmetic comments

kev-inn committed:
Resolve most PR comments

kev-inn committed:
Improve `pre_setup.py` and method method definitions to be similar to numpy

Calling `pre_setup.py` after building SystemDS (`mvn package -DskipTests`) will copy `lib` directory and `systemds-*-SNAPSHOT.jar`, which allows you to run the tests.

Calling `pre_setup.py`, `setup.py sdist` and `post_setup.py`, will create our package and move it to `target`, which can then be uploaded to pip (I will look into the specific steps to publish it).

kev-inn commented:
Btw I specified `author_email='damslab@mlist.tugraz.at'`, I do not know if this is the address we want to be contacted by.

kev-inn committed:
Fix setup for `sdist` (provide jar and so on) and add licenses

kev-inn commented:
Documentation will be added soon.

Baunsgaard commented on src/main/python/NOTICE:
```diff
@@ -0,0 +1,8 @@
+SystemDS
+Copyright 2018 Graz University of Technology
```
update

kev-inn replied:
should I replace with `[2018-2020]` or `2020`. The `NOTICE` file in the root dir should then also be updated I guess?

Baunsgaard replied:
I think that we can update the notice to say `[2018-2020]` and then write 2020 in these files.

Baunsgaard commented on src/main/python/pre_setup.py:
```diff
@@ -0,0 +1,44 @@
+    shutil.rmtree(java_dir_full_path, True)
+os.mkdir(java_dir_full_path)
+root_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.getcwd())))
+JAR_FILE_NAMES = ['systemds-*-SNAPSHOT.jar', 'systemds-*.jar', 'systemds-*-SNAPSHOT-extra', 'systemds-*-extra.jar']
```
Maybe not take any SystemDS, but the latest one?

kev-inn replied:
This would require us to each time we want to distribute a new package to make changes to this file, instead of just compiling with maven and executing `pre_setup.py`, `setup.py` and `post_steup.py`. I would prefer the second option.

Baunsgaard replied:
true, but it would ensure that the interface it compatible with the system. If you don't like it, then do not make the change :)

Baunsgaard commented on src/main/python/setup.py:
```diff
@@ -0,0 +1,100 @@
+ARTIFACT_VERSION_SHORT = ARTIFACT_VERSION.split("-")[0]
+
+numpy_version = '1.8.2'
+py4j_version = '0.10.0'
```
You might want to make a requirements.txt also, for easy standard installation.

https://stackoverflow.com/questions/7225900/how-to-install-packages-using-pip-according-to-the-requirements-txt-file-from-a#15593865

kev-inn replied:
Setuputils creates the requirements file automatically, that`s what the `setup.py` is for.

Baunsgaard reviewed (commented):
minor things for the changes.

kev-inn committed:
Fix issues and changes to setup

kev-inn committed:
Move all files on pre_setup.py to libs

Simplify code and prepare for prep release 0.2

kev-inn commented:
Note: the building of the packaging needs to be work in combination with @corepointer 's efforts towards 0.2.0, and the reworked `pom.xml`, therefore we could either accept one of the PR's and merge it into the other one, or merge both and create a whole new PR for packaging (I could delete the packaging files `setup.py`, `pre_setup.py`... for now.)

Baunsgaard commented:
I'm Merging this into master, with some slight modifications.
While doing so i have problems executing the tests, that all result in resource warnings.

`/usr/lib/python3.7/subprocess.py:883: ResourceWarning: subprocess 32620 is still running`

what was the fix again @kev-inn ?

also we have to talk about supported versions, since trying it with anything older than python 3.7 doesn't work, and as far as i know python 3.6 is the default ATM.

kev-inn commented:
@Baunsgaard I will look into the warning.

Yes, I depend on 3.7 for https://www.python.org/dev/peps/pep-0563/ . There is an old way of doing the delayed evaluation, I will switch to that.

kev-inn commented:
@Baunsgaard concerning the warning: It happens because we start a subprocess for the SystemDS execution, but it is hard to determine when we want to close that started process, because we never know that no more SystemDS execution is necessary. I tried to do it `atexit`, but py4j doesn't appreciate that. I will look into it, but if I feel like a `Context` is necessary I might postpone it for now.
Strangely if I execute the scripts in intellij I do not get the warning, I will take a closer look, but focus on other things for now.
